controller/server_status.py

```diff
@@ -0,0 +1,316 @@
+            output = {}
+            for status in statuses:
+                if status.completed_successfully or status.running_ready:
+                    output[status.container_name] = "ready"
+                elif status.failed:
+                    output[status.container_name] = "failed"
+                elif status.running:
+                    output[status.container_name] = "executing"
+                else:
+                    output[status.container_name] = "waiting"
+            return output
+
+        return {
+            "init": _get_summary(self.init_statuses),
+            "regular": _get_summary(self.statuses),
+        }
+
+    @property
+    def is_unschedulable(self) -> bool:
+        """Determines is a server pod is unschedulable."""
+        return (
+            self.pod_phase == K8sPodPhaseEnum.pending
+            and len(self.pod_conditions) >= 1
+            and self.pod_conditions[0].reason == "Unschedulable"
+            # NOTE: every pod is initially unschedulable until a PV is provisioned
+            # therefore to avoid "flashing" this state when a sessions starts this case is ignored
+            and isinstance(self.pod_conditions[0].message, str)
+            and "persistentvolumeclaim" not in self.pod_conditions[0].message.lower()
+        )
+
+    def server_url_is_eventually_responsive(self, timeout_seconds: int = 5) -> bool:
+        start = datetime.now()
+        while(True):
+            try:
+                res = requests.get(self.server_url, timeout=1)
+            except (requests.exceptions.RequestException, TimeoutError) as err:
+                logging.warning(
+                    f"Could not check session full URL {self.server_url} because error: {type(err)}"
+                )
+            else:
+                if res.status_code >= 200 and res.status_code < 400:
+                    return True
+            if (datetime.now() - start).total_seconds() > timeout_seconds:
+                return False
+            else:
+                sleep(1)
+                continue
+
+    @property
+    def overall_status(self) -> ServerStatusEnum:
+        """Get the status of the jupyterserver."""
+        if self.deletion_timestamp:
+            return ServerStatusEnum.Stopping
+        if self.is_unschedulable:
+            return ServerStatusEnum.Failed
+        failed_statuses = []
+        failed_init_statuses = []
+        ok_statuses = []
+        ok_init_statuses = []
+        for status in self.statuses:
+            if status.failed:
+                failed_statuses.append(status)
+            elif status.running_ready or status.completed_successfully:
+                ok_statuses.append(status)
+        for status in self.init_statuses:
+            if status.failed:
+                failed_init_statuses.append(status)
+            elif status.running_ready or status.completed_successfully:
+                ok_init_statuses.append(status)
```

Comment:
The list here is not actually used, it could just be ints that get incremented. Also, could `self.statuses` and `self.init_statuses` be chained/combined and just a single for loop used? And then below it'd just be something like

`and num_ok_statuses == len(self.statuses)+len(self.init_statuses)` ? Actually you could just count total number of statuses in the for loop and not use `len` at all.

Reply:
@Panaetius I think I originally had some other plans with those and that is why I had the arrays. Anyhow you are right all that stuff is not needed.